### ISSUE TITLE
False positive with Rule S1858 and URI's

### ISSUE THREAD

**Issue by JuanZamudioGBM** [SonarSource/sonar-dotnet#3665]
### Description

I have a method where I receive a string, in one case I want to concatenate to Uris, so I do it like this;
`uri1.ToString() + uri2.ToString()`

With that I get a warning for the rule S1858, if I remove the ToString() I get a compiler error "Operator '+' cannot be applied to operands of type 'Uri' and 'Uri'"

### Repro steps
```
using System;

    class Program
    {
        static void Main()
        {
            Uri uri1 = new Uri("http://localhost/");
            Uri uri2 = new Uri("resource/1", UriKind.Relative);
            Test(uri1.ToString() + uri2.ToString());
        }

        static void Test(string something)
        {
            Console.WriteLine(something);
        }
    }
```

### Expected behavior
No warnig

### Actual behavior
Warning generated

### Known workarounds
None

### Related information

* SonarAnalyzer.CSharp 8.14.0.22654
* 16.7.6
* dotnet 3.1
* Windows 10 2024

**Comment by pavel-mikula-sonarsource** [comment]
Hi @JuanZamudioGBM ,

Thank you for reporting this case. One of those two occurrences is a False Positive. You can do this
```
Test(uri1.ToString() + uri2);
```
to do what the rule intends to do. You can also use 
```
Test(new Uri(uri1, uri2))
```
to combine two Uris together.

**Comment by pavel-mikula-sonarsource** [comment]
Internal ticket [NET-1949](https://sonarsource.atlassian.net/browse/NET-1949)

[NET-1949]: https://sonarsource.atlassian.net/browse/NET-1949?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
